install.sh: add supervisor support

syuu1228 opened the pull request:
Bring supervisor support from dist/docker to install.sh, make it
installable from relocatable package.
This enables to use supervisor with nonroot / offline environment,
and also make relocatable package able to run in Docker environment.

Related #8849

syuu1228 commented:
Limitations:
 - It does not configure ulimits, user has responsibility to configure it in /etc/security/limits.conf. I did not added it because nonroot environment does not able to raise parameters than system default parameter, and system default parameter is not enough at all, at least on CentOS7.
 We may able to add auto-config of ulimits just for root privilege users.
 - Currently it still install systemd units, both on root mode and nonroot mode.
 - It does not install supervisor itself. On nonroot mode, user will need to install it via `pip` and prepare supervisord.conf.

yaronkaikov commented:
@syuu1228 It's not working for me. i still getting the following error
```
+ buildah run alpine-working-container-23 bash -x ./install.sh --nonroot --python3 /home/yaronkaikov/scylladb/python3/bin/python3
+ set -e
+ '[' -z '5.1.4(1)-release' ']'
+ root=/
+ housekeeping=false
+ nonroot=false
+ supervisor=false
+ supervisor_log_to_stdout=false
+ '[' 3 -gt 0 ']'
+ case "$1" in
+ nonroot=true
+ shift 1
+ '[' 2 -gt 0 ']'
+ case "$1" in
+ python3=/home/yaronkaikov/scylladb/python3/bin/python3
+ shift 2
+ '[' 0 -gt 0 ']'
+ '[' '!' -d /run/systemd/system/ ']'
+ false
+ echo 'systemd is not detected, unsupported distribution.'
systemd is not detected, unsupported distribution.
+ exit 1
error while running runtime: exit status 1
ERRO[0000] exit status 1 
```

avikivity commented:
> Limitations:
> 
> * It does not configure ulimits, user has responsibility to configure it in /etc/security/limits.conf. I did not added it because nonroot environment does not able to raise parameters than system default parameter, and system default parameter is not enough at all, at least on CentOS7.

On modern systemd, it raises the hard limit to ~500000, and ec60f44b6493abdde7bc765d7925fe2c8df3ec02 moves the soft limit to the hard limit. So above CentOS 7 we'd be fine.

>   We may able to add auto-config of ulimits just for root privilege users.

Let's see if we can get away without it.

> * Currently it still install systemd units, both on root mode and nonroot mode.

If we don't enable the units, that's fine, I think.

> * It does not install supervisor itself. On nonroot mode, user will need to install it via `pip` and prepare supervisord.conf.

That's fine too.

avikivity commented on dist/common/supervisor/scylla-jmx.sh:
```diff
@@ -0,0 +1,7 @@
+
+. "$sysconfdir"/scylla-jmx
+
+execsudo "$scylladir"/jmx/scylla-jmx $SCYLLA_JMX_PORT $SCYLLA_API_PORT $SCYLLA_API_ADDR $SCYLLA_JMX_ADDR $SCYLLA_JMX_FILE $SCYLLA_JMX_LOCAL $SCYLLA_JMX_REMOTE $SCYLLA_JMX_DEBUG
```
`exec sudo`?

avikivity replied:
and why sudo?

syuu1228 replied:
> `exec sudo`?

it's
```
execsudo() {
    if is_nonroot; then
        exec "$@"
    else
        exec sudo -u scylla -g scylla "$@"
    fi
}
```
in scylla_util.sh.

Why I added it is, because scylla_prepare and scylla has to run in different user.
scylla_prepare is run in root, but scylla is run in scylla.
But it's only scylla-server problem, we don't need to do `exec sudo -u scylla` on other services such as scylla-jmx, so I will drop it and replace with `user=scylla` directive on supervisor config file.


About scylla-server, on very early version of scylla we did opposite:

scylla-server.service:
```
ExecStartPre=/usr/bin/sudo -E /usr/lib/scylla/scylla_prepare
ExecStart=/usr/bin/scylla $SCYLLA_ARGS $SEASTAR_IO $DEV_MODE
User=scylla
```

sudoers.d/scylla:
```
scylla ALL=(ALL) NOPASSWD: /usr/lib/scylla/scylla_prepare,/usr/lib/scylla/scylla_stop,/usr/lib/scylla/scylla_io_setup,/usr/lib/scylla/scylla-ami/scylla_ami_setup
```

But it wasn't enough, we also needed to rewrite sudoers:
```
%if 0%{?rhel}
sed -e "s/Defaults    requiretty/#Defaults    requiretty/" /etc/sudoers > /tmp/sudoers
cp /tmp/sudoers /etc/sudoers
rm /tmp/sudoers
%endif
```

It seems like to run `sudo -u scylla` in root is much easier than opposite way, since there are restrictions to become root and it should be doing carefully.

avikivity replied:
Ah, ok. But sudo suggests that it will run as root, suggest calling it run_as_scylla_user to be super clear.

syuu1228 commented:
Changelog:

Dropped `execsudo` on scylla-jmx, scylla-node-exporter, use `user` directive instead.
Dropped scylla-housekeeping service, since it seems like does not work for relocatable package.
Add Debian variant support, since it uses different directory to store supervisor config files by default.
Stop running systemctl when --supervisor specified, since it causes error when it run in Docker, since there is no systemd service inside container.
Rebased with master.

syuu1228 commented:
Also added submodule changes:
 - https://github.com/scylladb/scylla-jmx/pull/171
 - https://github.com/scylladb/scylla-tools-java/pull/262
 - https://github.com/scylladb/scylla-python3/pull/21

syuu1228 commented:
It's out side of scylla repository, but implemented test code to run supervisor in Docker container:
https://github.com/syuu1228/scylla-supervisor-test

syuu1228 commented:
> @syuu1228 It's not working for me. i still getting the following error

Fixed bugs around systemd check and also added support alpine.
Please try this, a Docker image to run supervisor on alpine:
https://github.com/syuu1228/scylla-supervisor-test/tree/main/alpine

syuu1228 commented:
@yaronkaikov Uploaded locally built unified package which able to construct alpine Docker image, if it doesn't work on your side please try this one:
https://drive.google.com/file/d/1cv5SG9CMka5XSXWxCG4vPlw_or54WERY/view?usp=sharing

syuu1228 commented:
alpine is busybox based minimal distro, but our installer didn't work correctly pure-busybox only environment, so I had to install normal linux CLI tools outside of busybox:
```
sudo tar bash coreutils grep shadow
```
https://github.com/syuu1228/scylla-supervisor-test/blob/main/alpine/Dockerfile

If we want to diet container image, we need to fix it too.

yaronkaikov commented:
> alpine is busybox based minimal distro, but our installer didn't work correctly pure-busybox only environment, so I had to install normal linux CLI tools outside of busybox:
> 
> ```
> sudo tar bash coreutils grep shadow
> ```
> 
> https://github.com/syuu1228/scylla-supervisor-test/blob/main/alpine/Dockerfile
> 
> If we want to diet container image, we need to fix it too.

@syuu1228 Looks like we should. currently our docker image is about 800Mb (check version 4.4.3) with `alpine` is about 5Gb

yaronkaikov commented:
> @yaronkaikov Uploaded locally built unified package which able to construct alpine Docker image, if it doesn't work on your side please try this one:
> https://drive.google.com/file/d/1cv5SG9CMka5XSXWxCG4vPlw_or54WERY/view?usp=sharing

It's working:-)

syuu1228 commented:
> @syuu1228 Looks like we should. currently our docker image is about 800Mb (check version 4.4.3) with `alpine` is about 5Gb

There are two (or more) reasons why it's big:
First reason is I didn't removed unified package & extracted directory.
I able to shrink half of image size just removing them:
```
FROM alpine:latest
RUN apk add --no-cache supervisor openjdk8-jre-base sudo tar bash coreutils grep shadow
#COPY scylla-unified-package.tar.gz /
RUN mkdir /unified && wget http://192.168.100.253:8000/scylla-unified-package-4.6.dev.0.20210626.f33e7204a.tar.gz && tar xpf scylla-unified-package-4.6.dev.0.20210626.f33e7204a.tar.gz -C /unified && cd /unified && ./install.sh --supervisor --supervisor-log-to-stdout && rm -rf /unified /scylla-unified-package-*.tar.gz
RUN scylla_dev_mode_setup --developer-mode 1
ENTRYPOINT ["supervisord", "-n"]
```
But it still 2.3GB:
```
[syuu@monster ~]$ docker images scylla-unified-test-alpine:5
REPOSITORY                   TAG       IMAGE ID       CREATED         SIZE
scylla-unified-test-alpine   5         48f90c22c079   5 minutes ago   2.3GB
```

Second reason is, it seems scylla binary in rerocatable package is unstripped, scylla binary eats 1.8GB:
```
bash-5.1# du -sch /opt/scylladb/*
0       /opt/scylladb/SCYLLA-OFFLINE-FILE
328K    /opt/scylladb/api
48K     /opt/scylladb/bin
9.4M    /opt/scylladb/jmx
1.9G    /opt/scylladb/libexec
69M     /opt/scylladb/libreloc
19M     /opt/scylladb/node_exporter
66M     /opt/scylladb/python3
684K    /opt/scylladb/scripts
80K     /opt/scylladb/scyllatop
34M     /opt/scylladb/share
16K     /opt/scylladb/supervisor
4.0M    /opt/scylladb/swagger-ui
4.0K    /opt/scylladb/uninstall.sh
2.1G    total
bash-5.1# du -sch /opt/scylladb/libexec/*
524K    /opt/scylladb/libexec/ethtool
704K    /opt/scylladb/libexec/gawk
100K    /opt/scylladb/libexec/gzip
52K     /opt/scylladb/libexec/hwloc-calc
32K     /opt/scylladb/libexec/hwloc-distrib
84K     /opt/scylladb/libexec/ifconfig
64M     /opt/scylladb/libexec/iotune
120K    /opt/scylladb/libexec/lsblk
88K     /opt/scylladb/libexec/lscpu
168K    /opt/scylladb/libexec/netstat
160K    /opt/scylladb/libexec/patchelf
1.8G    /opt/scylladb/libexec/scylla
4.0K    /opt/scylladb/libexec/ubsan-suppressions.supp
1.9G    total
```

avikivity commented:
We can strip it post install (maybe add install.sh --strip).

avikivity commented:
> We can strip it post install (maybe add install.sh --strip).

Not so easy, `strip` is part of binutils, and we don't want to depend on it or carry `strip` with us.

syuu1228 commented:
@avikivity Can we strip debug symbol during building scylla-package.tar.gz or scylla-unified-package.tar.gz, and add debug symbol package separately just like rpm/deb do?

avikivity commented:
I think it's a good idea, but will take some work. We have several options:

1. make scylla-package.tar.gz contain separate binary and debuginfo, add option to install.sh to skip installing debuginfo
2. have scylla-package-tar.gz and scylla-package-debuginfo.tar.gz. This allows for smaller downloads but complicates building rpm/deb (they need two sources).

I think 1 is fine. The tarball is not used very commonly so it's okay if it's larger.

Yet another option is to keep using a package. It seems alpine doesn't buy us much for reducing the docker image size, since we'll need to install coreutils and a lot of other stuff to make it work.

penberg commented:
@avikivity I understood that the motivation for `supervisord` support was to run Scylla on Alpine Linux? Now that we're moving Docker image to Ubuntu, do we still want these to be merged?

avikivity commented:
Yes, there's less motivation now. It can be used to run rootless on versions of systemd that are too old to have --user, but that's a tiny use case.

syuu1228 commented:
@penberg @avikivity I know it's tiny use case but can we proceed to merge this, to use CentOS7 nonroot install?
Also it could be useful for non-systemd environment such as gentoo linux or inside of container.
